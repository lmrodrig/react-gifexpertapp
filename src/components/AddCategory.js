import React, { useState } from 'react'
import PropTypes from 'prop-types';   
 

export const AddCategory = ({setCategories}) => {
    //el estado default se debe de declarar sino sera undefined
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) =>{
        setInputValue(e.target.value);
        //console.log('HandleInput changed...!');

    }

    const handleSubmit = (e) =>{
        //para quedarnos en la misma ruta sin agregar /? al path
        e.preventDefault();
        if(inputValue.trim().length > 2){
            setCategories( cats => [inputValue,...cats]);
            setInputValue('');
        }
        //console.log('Se llamo el handleSubmit..');
        
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
           <input
                type='text'
                value={inputValue}
                onChange={handleInputChange}
           />
        </form>
    )
    
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired 
    
}