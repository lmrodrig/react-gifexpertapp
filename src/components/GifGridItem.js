import React from 'react'

export const GifGridItem = ( { title, url}) => {

    //react toma className como class de CSS
    return (
        <div className='card animate__animated animate__fadeIn'>
            <img src={url} alt={title}></img>
            <p>{title}</p>
        </div>
    )
}
