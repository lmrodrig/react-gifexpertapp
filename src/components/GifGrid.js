import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types';  

import { useFetchGifs } from "../hooks/UseFetchGifs";
import { GifGridItem } from './GifGridItem';
//import { getGifs } from '../helpers/GetGifs';

export const GifGrid = ({ category}) => {

    
    //const [images, setImages] = useState([]);
    //Este Hook dice que solo se debe de ejecutar 1 sola vez
    //cuando se inicializa el componente.
    // useEffect(() => {
    //     //haremos una peticion HTTP con fetch
    //     getGifs(category)
    //     .then( imgs => setImages(imgs));
    // }, [category]); // Aqui solo le indica que si la categoria cambia se lanza otra vez
    //                 // la llamada a getGifs 
        
    // en el uso del map se pide tener una key unica en el GifGridItem
    // por eso se pone como propiedad del elemento
    // se puede confundir que es un parametro, pero no lo es. 

    //usando custom hooks
    const { data:images, loading } = useFetchGifs( category );

    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> { category } </h3>

            { loading && <p className="animate__animated animate__flash">Loading</p> }

            <div className="card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}