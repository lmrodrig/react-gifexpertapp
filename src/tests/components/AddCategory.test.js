import '@testing-library/jest-dom'
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

//const setCat = () => {}; jest.fn hace lo mismo para simular una funcion
const setCat = jest.fn();
let wrapper = shallow( <AddCategory setCategories={ setCat} />); 

beforeEach( () => {
    jest.clearAllMocks();
    wrapper = shallow( <AddCategory setCategories={ setCat} />); 
});

describe('Pruebas en <AddCategory /> component', () => {
    // Modo corto de enviar una funcion como parametro
    //const wrapper = shallow( <AddCategory setCategories={ ()=>{}} />);
       
    test('Debe de mostrar < AddCategory /> correctamente', () => {
        
        //se crea un snapshot en la carpeta de test
        expect (wrapper).toMatchSnapshot();
    })

    test('Debe de cambiar el valor del Input box', () => {
       const input = wrapper.find('input');
       const valor = 'Hola Mundo';

       //estamos simulando el parametro de este handler: setInputValue(e.target.value);
       input.simulate('change', {
           target : {
               value : valor
           }
       });

       expect( wrapper.find('p').text().trim()).toBe(valor);

    });

    test('No debe de postear con onSubmit ', () => {
      wrapper.find('form').simulate('submit',{
        preventDefault(){}
      });
    });
    //como input no se envio (esta vacio) no es >2 no se debe de llamar setCategories
    expect( setCat ).not.toHaveBeenCalled();

    test('Debe de llamar a la fn setCategories y limpiar el input Box', () => {
    
        const input = wrapper.find('input');
        const valor = 'Hola Mundo';
        //1.- Simular el input change
        input.simulate('change', {
            target : {
                value : valor
            }
        });

        //2.- simular el submit
        wrapper.find('form').simulate('submit', {
              preventDefault(){}
        })

        //3.- setCategories se debe haber llamado
        expect(setCat).toHaveBeenCalled();

        //4.- el valor del input box debe ser ''
        expect(wrapper.find('input').prop('value')).toBe('');

    });
    
    
})