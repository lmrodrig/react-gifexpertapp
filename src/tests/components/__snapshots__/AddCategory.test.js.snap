// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> component Debe de mostrar < AddCategory /> correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
